Remove unnecessary typecasts and add promotion to support Dual numbers

dennisYatunin opened the pull request:
<!--- THESE LINES ARE COMMENTED -->
## Purpose 
<!--- One sentence to describe the purpose of this PR, refer to any linked issues:
#14 -- this will link to issue 14
Closes #2 -- this will automatically close issue 2 on PR merge
-->
This PR is a followup to CliMA/RootSolvers.jl#50, CliMA/Thermodynamics.jl#199, and CliMA/Thermodynamics.jl#200. Together with those changes, this should allow us to pass `ForwardDiff.Dual` numbers through the ClimaAtmos implicit tendency.

## Content
<!---  specific tasks that are currently complete 
- Solution implemented
-->
- Remove unused `FT` type variables from method definitions
- Drop typecasts to `FT` where the argument is derived from the parameter set
- For typecasts to `FT` where the argument is a `Float64` literal (`0.01` or `Inf`), infer `FT` from the parameter set instead of the surface conditions (which might contain a mixture of ordinary numbers and `Dual` numbers)
- Add a restriction to the definition of `SurfaceFluxesParameters` to ensure that the parameter set has an internally consistent `FT` type
- Add a constructor to `SurfaceFluxConditions` that promotes all of its arguments to a common type (for mixtures of ordinary numbers and `Dual` numbers, this will promote everything to `Dual`s)
- Fix a typo: replace `Δz(sc) ./ ζₛ` with `Δz(sc) / ζₛ` (both values are scalars, so we do not need to broadcast here)
  - Fixes #144
- Fix a type instability: replace `π_group⁻¹ = 1 / π_group; return π_group⁻¹ * von_karman_const / Σterms` with `return von_karman_const / (π_group * Σterms)` (since `π_group` is 1, computing `π_group⁻¹` results in a `Float64` type instability)
- Bump the minor version number

<!---
Review checklist

I have:
- followed the codebase contribution guide: https://clima.github.io/ClimateMachine.jl/latest/Contributing/
- followed the style guide: https://clima.github.io/ClimateMachine.jl/latest/DevDocs/CodeStyle/
- followed the documentation policy: https://github.com/CliMA/policies/wiki/Documentation-Policy
- checked that this PR does not duplicate an open PR.

In the Content, I have included 
- relevant unit tests, and integration tests, 
- appropriate docstrings on all functions, structs, and modules, and included relevant documentation.

-->

----
- [x] I have read and checked the items on the review checklist.

charleskawczynski commented on src/SurfaceFluxes.jl:
```diff
@@ -427,19 +429,19 @@ function obukhov_length(
         κ = SFP.von_karman_const(param_set)
         Cd = κ^2 / (log(ϵₘ)^2) * (1 - ψₘ / log(ϵₘ))^(-2)
         Ch = κ^2 / (Pr₀ * log(ϵₘ) * log(ϵₕ)) * (1 - ψₘ / log(ϵₘ))^(-1) * (1 - ψₕ / Pr₀ / log(ϵₕ))^(-1)
-        return non_zero(Δz(sc) ./ ζₛ)
```
This might have actually been the gpu issue. Speaking of which, maybe we can test this in ClimaAtmos (https://github.com/CliMA/SurfaceFluxes.jl/issues/144). Actually, it could be the use of kwargs below. I'll kick off a job with this branch to try it out.

dennisYatunin replied:
Yes, this appears to have been the GPU issue. I've reverted the change described in that issue, and the GPU test is still passing. I just started a ClimaAtmos job to try this out.

charleskawczynski commented on src/SurfaceFluxes.jl:
```diff
@@ -616,18 +616,16 @@ function heat_exchange_coefficient(
     ΔDSEᵥ = DSEᵥ_in - DSEᵥ_sfc
     z0_b = z0(sc, UF.HeatTransport())
     z0_m = z0(sc, UF.MomentumTransport())
-    if isapprox(ΔDSEᵥ, FT(0); atol = tol_neutral)
```
Not an objection, but I am wondering if this is equivalent. (I could see `isapprox` behaving differently between floats and Ints)

dennisYatunin replied:
Looking at the output of `code_llvm` and `code_native`, it seems like the overall code is the same, but using an`Int` adds a few promotion-related instructions. We can also rewrite this as `abs(ΔDSEᵥ) <= tol_neutral` to use even fewer instructions.

charleskawczynski replied:
I could see the gpu compiler liking `abs(ΔDSEᵥ) <= tol_neutral` more than `isapprox` tbh

charleskawczynski reviewed (approved):
Overall these changes look fine to me. I'm going to add a ClimaAtmos job to make sure that this will run okay on the GPU. We should really dig into what the issue was in https://github.com/CliMA/SurfaceFluxes.jl/issues/144. Until then, I think we'll need to run integration tests to make sure that we're not breaking upstream code.

dennisYatunin commented:
I fixed an a `Float64` type instability, and the tests are now passing. I also reverted the change described in #144, since it was only necessary because of a typo that I removed in this PR. Currently trying out in ClimaAtmos here: https://buildkite.com/clima/climaatmos-ci/builds/16829

dennisYatunin commented:
The ClimaAtmos examples all ran correctly. There were some minor MSE table changes due to rearranged floating-point operations, but the plots look identical. Merging this in.
